llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp

```diff
@@ -10984,6 +11087,12 @@ unsigned ARMAsmParser::checkTargetMatchPredicate(MCInst &Inst) {
     if (!hasV8Ops() && (Inst.getOperand(0).getReg() == ARM::SP))
       return Match_RequiresV8;
     break;
+  case ARM::tMUL:
+    // The second source operand must be the same register as the destination
+    // operand.
+    if (Inst.getOperand(0).getReg() != Inst.getOperand(3).getReg())
+      return Match_RequiresDestinationRegisterMatchASourceRegister;
```

Comment:
This should be diagnosed by built-in tied operand checking. Does it not do that for some reason?

Reply:
There's weirdness in this case. I can't quite remember why. I think because of `cvtThumbMultiply` being the custom match class that tries to swap `Rn` and `Rm` if needed to get them to match the normal method for this check doesn't work? 

Without this the matching of `muls r1, r2, r3` hits the error at llvm/lib/Target/ARM/AsmParser/ARMAsmParser.cpp:8036 when validating the narrow version so doesn't attempt to match the wide version.

Reply:
Hmm yeah I can't see where in `build/lib/Target/ARM/ARMGenAsmMatcher.inc` such operand restrictions would be checked. Is this supported for instruction selection at all do you know?

Reply:
It appears the code that checks the constraints is not emitted when ReportMultipleNearMisses is true:
```C++
  if (!ReportMultipleNearMisses)
    emitAsmTiedOperandConstraints(Target, Info, OS, HasOptionalOperands);
```
There is a comment elsewhere:
```
      // FIXME: The tied operands checking is not yet integrated with the
      // framework for reporting multiple near misses.
```
In any case, it seems that if an instruction has a custom converter, the built-in tied operands checking is not performed. 

PS
There is `Match_InvalidTiedOperand` that would be used if tied operand checking was emitted.

Reply:
Ah great, thank you. That's a bit of a pain but for now I will add a `FIXME` explaining. I'll switch to using `Match_InvalidTiedOperand` also.

Reply:
Done